File: python/dssatutils/soil_slga.py
# Soil source: Soil and Landscape Grid of Australia (SLGA).

import os

import numpy as np
import pandas as pd

from .soil_raster_common import (
    DEPTHS, add_physics_texture, coords, find_raster, normalize_bd,
    sample_raster, write_mapping, write_profiles,
)


def process_soils_slga(
    grid_points,
    slga_raster_dir: str,
    output_csv_path: str,
    output_sol_dir: str,
    id_col: str = "ID",
    lat_col: str = "LAT",
    long_col: str = "LONG",
    depth_specs=None,
    source_name: str = "Soil and Landscape Grid of Australia",
    source_tag: str = "SLGA local rasters",
) -> None:
    """Build DSSAT .SOL files from local per-property/per-depth GeoTIFF/VRT rasters.

    *depth_specs* is a list of ``(depth_token, depth_bottom_cm, depth_center_cm)``
    tuples; it defaults to the SoilGrids standard depths and can be overridden for
    products with different layering (e.g. WISE30sec). Rasters are matched by a
    property token (``sand``/``clay``/``silt``/``bulk``/``organic carbon``) AND the
    depth token appearing in the filename.
    """
    if not os.path.isdir(slga_raster_dir):
        raise FileNotFoundError(f"SLGA raster directory not found: {slga_raster_dir}")
    if depth_specs is None:
        depth_specs = DEPTHS
    ids, lats, lons = coords(grid_points, id_col, lat_col, long_col)
    rows = []
    for dtoken, dbot, dctr in depth_specs:
        sand = sample_raster(find_raster(slga_raster_dir, ["sand"], dtoken), lats, lons)
        clay = sample_raster(find_raster(slga_raster_dir, ["clay"], dtoken), lats, lons)
        silt_path = find_raster(slga_raster_dir, ["silt"], dtoken)
        silt = sample_raster(silt_path, lats, lons) if silt_path else 100.0 - sand - clay
        bd = normalize_bd(sample_raster(find_raster(slga_raster_dir, ["bulk"], dtoken)
                                        or find_raster(slga_raster_dir, ["bd"], dtoken), lats, lons))
        soc = sample_raster(find_raster(slga_raster_dir, ["organic", "carbon"], dtoken)
                            or find_raster(slga_raster_dir, ["soc"], dtoken), lats, lons)
        soc_pct = np.where(soc > 20, soc / 10.0, soc)
        for i, pid in enumerate(ids):
            rows.append({
                "ID": pid, "latitude": lats[i], "longitude": lons[i],
                "depth_bottom": dbot, "depth_center": dctr,
                "sand": sand[i], "clay": clay[i], "silt": silt[i],
                "bdod": bd[i], "soc_pct": soc_pct[i], "cfvo": 0.0,
            })
    df = pd.DataFrame(rows).dropna(subset=["sand", "clay", "bdod", "soc_pct"])
    if df.empty:
        raise RuntimeError("No usable SLGA soil data extracted. Check raster names and coordinates.")
    df = add_physics_texture(df)
    write_mapping(ids, output_csv_path)
    write_profiles(df, output_sol_dir, source_name, source_tag)

File: R/soil_slga.R
process_soils_slga <- function(grid_points, slga_raster_dir, output_csv_path,
                               output_sol_dir, id_col = "ID",
                               lat_col = "LAT", long_col = "LONG",
                               depth_specs = NULL,
                               source_name = "Soil and Landscape Grid of Australia",
                               source_tag = "SLGA local rasters") {
  if (!dir.exists(slga_raster_dir)) stop(sprintf("SLGA raster directory not found: %s", slga_raster_dir))
  if (is.null(depth_specs)) depth_specs <- SOIL_RASTER_DEPTHS
  pts <- sf::st_transform(grid_points, 4326)
  ids <- as.character(sf::st_drop_geometry(pts)[[id_col]])
  xy <- sf::st_coordinates(pts); lats <- xy[, 2]; lons <- xy[, 1]
  pts_vect <- terra::vect(pts)
  rows <- list()
  for (j in seq_len(nrow(depth_specs))) {
    d <- depth_specs[j, ]
    sand <- soil_sample_raster(soil_find_raster(slga_raster_dir, "sand", d$token), pts_vect)
    clay <- soil_sample_raster(soil_find_raster(slga_raster_dir, "clay", d$token), pts_vect)
    silt_p <- soil_find_raster(slga_raster_dir, "silt", d$token)
    silt <- if (!is.na(silt_p)) soil_sample_raster(silt_p, pts_vect) else 100 - sand - clay
    bd <- soil_sample_raster(soil_find_raster(slga_raster_dir, c("bulk"), d$token), pts_vect)
    bd <- ifelse(bd > 10, bd / 100, bd)
    soc <- soil_sample_raster(soil_find_raster(slga_raster_dir, c("organic", "carbon"), d$token), pts_vect)
    soc_pct <- ifelse(soc > 20, soc / 10, soc)
    rows[[j]] <- data.frame(ID = ids, latitude = lats, longitude = lons,
                            depth_bottom = d$bottom, depth_center = d$center,
                            sand = sand, clay = clay, silt = silt,
                            bdod = bd, soc_pct = soc_pct, cfvo = 0)
  }
  df <- do.call(rbind, rows)
  df <- df[complete.cases(df[, c("sand", "clay", "bdod", "soc_pct")]), ]
  if (!nrow(df)) stop("No usable SLGA soil data extracted. Check raster names and coordinates.")
  df <- soil_add_physics(df)
  soil_write_mapping(ids, output_csv_path)
  soil_write_profiles(df, output_sol_dir, source_name, source_tag)
}
